docs/demo/data/processed/model_extensions.py

```diff
@@ -0,0 +1,307 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Dec 19 10:50:55 2023
+
+@author: bdobson
+"""
+from wsimod.core import constants
+import sys
+import os
+from tqdm import tqdm
+from math import log10
+from wsimod.nodes.land import ImperviousSurface
+from wsimod.nodes.nodes import QueueTank, Tank, ResidenceTank
+
+def extensions(model):
+    # Apply customations
+    model.nodes['my_groundwater'].residence_time *= 2
+    
+    # Add dcorator to a reservoir
+    model.nodes['my_reservoir'].net_evaporation = model.nodes['my_reservoir'].empty_vqip()
+    model.nodes['my_reservoir'].mass_balance_out.append(lambda self=model.nodes['my_reservoir']: self.net_evaporation)
+    model.nodes['my_reservoir'].net_precipitation = model.nodes['my_reservoir'].empty_vqip()
+    model.nodes['my_reservoir'].mass_balance_in.append(lambda self=model.nodes['my_reservoir']: self.net_precipitation)
+
+    model.nodes['my_reservoir'].make_abstractions = wrapper_reservoir(model.nodes['my_reservoir'], model.nodes['my_reservoir'].make_abstractions)
+    
+    # Change model.run because new function has been introduced by the new node
+    model.run = wrapper_run(model)
+    
+def wrapper_reservoir(node, func):
+    def reservoir_functions_wrapper():
+        # Initialise mass balance VQIPs
+        vqip_out = node.empty_vqip()
+        vqip_in = node.empty_vqip()
+        
+        # Calculate net change
+        net_in = node.get_data_input('precipitation') - node.get_data_input('et0')
+        net_in *= node.tank.area
+        
+        if net_in > 0:
+            # Add precipitation
+            vqip_in = node.v_change_vqip(node.empty_vqip(), net_in)
+            _ = node.tank.push_storage(vqip_in, force = True)
+            
+        else:
+            # Remove evaporation
+            evap = node.tank.evaporate(-net_in)
+            vqip_out = node.v_change_vqip(vqip_out, evap)
+        
+        # Store in mass balance states
+        node.net_evaporation = vqip_out
+        node.net_precipitation = vqip_in
+        
+        node.satisfy_environmental()
+        # Call whatever else was going happen
+        return func()
+    return reservoir_functions_wrapper
+
+def wrapper_run(self):
```

Comment:
This is very much NOT sustainable. It requires from anyone willing to add its own functionality to copy the whole of the `model.run` function and add bits here and there to run the custom functionality. That is really complicated.

The underlying problem is that each node has its own interface, which means it is impossible to abstract their execution: you need to call each type of node separately and run the specific function for that node. So, if you add a new node, you need to add a new step in `run` to execute it.

Reply:
@dalonsoa The `run` function is a really tight balance between enabling customisation over orchestration vs getting something tidy. In principle we would just like every node to have a `run` function that is called during the model `run` timestep and that is that and very easy/sustainable. Though the whole point of having this more arduous orchestration is to facilitate customisation, e.g., note how some nodes have different functions that are called multiple times throughout the timestep within the `run` function - triggering one node, then another, and then the first node - is needed to capture many systems.

I definitely understand it is not particularly satisfying, but I can't say I came up with a better solution. One software I have used has 'before timestep', 'during timestep' and 'after timestep' functions (this inspired the architecture of the [`Surface`](https://imperialcollegelondon.github.io/wsi/reference-land/#wsimod.nodes.land.Surface) object with three function lists of `inflows`, `processes` and `outflows`). We could take a similar approach within the model's `run` function but a) it would require quite a lot of work implementing this behaviour for every node, and 2) I'm not convinced that many water systems would still probably require further customisation to capture more complicated interactions

Reply:
I guess a compromise solution could be that the user provides a list of <node_type> : <function_name> to the run function, and it can do the rest from there?

So to recreate the default `run` behaviour you would provide a list of dicts:
[{'FWTW' : 'treat_water'},
{'Demand' : 'create_demand'},
{'Land' : 'run'},
{'Groundwater' : 'infiltrate'},
{'Sewer' : 'make_discharge'},
{'Foul' : 'make_discharge'},
{'WWTW' : 'calculate_discharge'},
{'Groundwater' : 'distribute'},
{'River' : 'calculate_discharge'},
{'Reservoir' : 'make_abstractions'},
{'Land' : 'apply_irrigation'},
{'WWTW' : 'make_discharge'},
{'Catchment' : 'route'}]

And the run function would just iterate over the list and call functions accordingly